setools/terulequery.py

```diff
@@ -162,13 +172,17 @@ def results(self):
             #
             # Matching on target type
             #
-            if self.target and not match_indirect_regex(
+            if self.target and not \
+                    match_indirect_regex(
                     rule.target,
                     self.target,
                     self.target_indirect,
                     self.target_regex):
                 continue
 
+            if self.target_self and not rule.target == rule.source:
+                continue
```

Comment:
On further thought, I think this is too simplistic in cases of attributes, particularly for `allow attr attr:...`.

If you do `allow attr self:...` that expands the rule into `allow type1 type1:... allow type2 type2:...`. `allow attr attr` does not correspond to a source policy self rule. I'm inclined to say we don't want `allow attr attr` to match since that doesn't match a self rule as written in source, but I am open to arguments.

@stephensmalley @jwcart2 @jbrindle @doverride @cgzones @bauen1 any opinions?

Reply:
In my opinion, I think of sesearch as being more fundamentally about permission search than rule search.

If I run sesearch for "allow type1 self", and "allow attr attr" is in the policy, then the permission I searched for is present, even though the specific rule is not.  That's information I would like to get in that query.

Reply:
> If you do allow attr self:... that expands the rule into allow type1 type1:... allow type2 type2:.... allow attr attr does not correspond to a source policy self rule. I'm inclined to say we don't want allow attr attr to match since that doesn't match a self rule as written in source, but I am open to arguments.

I would argue that the `self` keyword is syntactic sugar to indicate the same target type as the source type.
`(allow a a (.))` and `(allow a self (.))` are the same, there for a search for allow rules with the source type `a` with either the same target type `a` or the syntactic sugar `self` should return both rules (maybe formatted accordingly to the query).

Reply:
> `(allow a a (.))` and `(allow a self (.))` are the same

This is true where `a` is a type, but not if it is an attribute.

Reply:
> I would argue that the self keyword is syntactic sugar to indicate the same target type as the source type.

self is not syntactic sugar, it is stored in the kernel policy and can express something not possible otherwise. Namely, when the source is an attribute, a self target means each type in the attribute as the source and itself as the target. For example, these are not equivalent:
allow domain domain:process fork;
allow domain self:process fork;

The first gives every domain process:fork for every domain, the latter gives every domain process:fork only on itself.

Reply:
Oh my brain just read what I wanted to read ...

> If you do allow attr self:... that expands the rule into allow type1 type1:... allow type2 type2:.... allow attr attr does not correspond to a source policy self rule. I'm inclined to say we don't want allow attr attr to match since that doesn't match a self rule as written in source, but I am open to arguments.

sesearch should output all rules that match `(allow a a (.))`, even in the form of typeattributes.
But (if possible, i.e. not if they have been expanded) the type attribute rules shouldn't use the `self` keyword when displaying the rule.

Reply:
> self is not syntactic sugar
Is that true for both type attributes and normal types ?

Reply:
I have to stick with my original thought: `--self` not match in `allow attr attr:...` statements.

Reply:
I'm having a hard time seeing the value of this change with that.  Do you have any example of a motivating use case where someone might use this feature with the behavior of --self not matching "allow attr attr"?

I get where you're coming from in terms of matching the kernel rules, but my motivating example to want this is a situation where I want to know if a given domain has certain self access. For example if I want to know which domains allow anonymous pipes, I might run:

`sesearch -A -c fifo_file --self`

If --self doesn't match "allow attr attr", then a policy with something like "allow domain domain:fifo_file rw_fifo_file_perms;" would not match.  That means that I can't assume that the above query gives me all domains with fifo_file access.  It's true that it doesn't match the policy rule perfectly (and that "allow domain self" is very different access), but I'm no longer sure what scenarios I would use this query in if there are potentially permissions that are specified using attributes.

Reply:
I'd like to get this merged, so lets make `--self` sensitive to `self.target_indirect`. Then the user can get either behavior.